hook-bootkit/main.go

```diff
@@ -71,11 +116,74 @@ func main() {
 		imageName = cfg.tinkWorkerImage
 	}
 	if imageName == "" {
-		// TODO(jacobweinstock): Don't panic, ever. This whole main function should ideally be a control loop that never exits.
-		// Just keep trying all the things until they work. Similar idea to controllers in Kubernetes. Doesn't need to be that heavy though.
-		panic("cannot pull image for tink-worker, 'docker_registry' and/or 'tink_worker_image' NOT specified in /proc/cmdline")
+		return fmt.Errorf("cannot pull image for tink-worker, 'docker_registry' and/or 'tink_worker_image' NOT specified in /proc/cmdline")
+	}
+
+	// Give time for Docker to start
+	// Alternatively we watch for the socket being created
+	log.Info("setuping up the Docker client")
+
+	os.Setenv("HTTP_PROXY", cfg.httpProxy)
+	os.Setenv("HTTPS_PROXY", cfg.httpsProxy)
+	os.Setenv("NO_PROXY", cfg.noProxy)
+	// Create Docker client with API (socket)
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+
+	log.Info("Pulling image", "imageName", imageName)
+	authConfig := registry.AuthConfig{
+		Username: cfg.username,
+		Password: strings.TrimSuffix(cfg.password, "\n"),
+	}
+
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return err
+	}
+
+	authStr := base64.URLEncoding.EncodeToString(encodedJSON)
+
+	pullOpts := types.ImagePullOptions{
+		RegistryAuth: authStr,
+	}
+	var out io.ReadCloser
+	imagePullOperation := func() error {
+		out, err = cli.ImagePull(ctx, imageName, pullOpts)
+		if err != nil {
+			log.Error(err, "Image pull failure", "imageName", imageName)
+			return err
+		}
+		return nil
+	}
+	if err = backoff.Retry(imagePullOperation, backoff.NewExponentialBackOff()); err != nil {
+		return err
+	}
+
+	if _, err = io.Copy(os.Stdout, out); err != nil {
+		return err
+	}
+
+	if err = out.Close(); err != nil {
+		log.Error(err, "error when closing image pull logs")
 	}
 
+	cs, err := cli.ContainerList(ctx, types.ContainerListOptions{All: true})
+	if err != nil {
+		return fmt.Errorf("error listing containers used for checking for an existing tink-worker container: %w", err)
+	}
+	for _, c := range cs {
+		for _, n := range c.Names {
+			if n == "/tink-worker" {
+				log.Info("Removing existing tink-worker container")
+				if err := cli.ContainerRemove(ctx, c.ID, types.ContainerRemoveOptions{Force: true}); err != nil {
+					return fmt.Errorf("error removing existing tink-worker container: %w", err)
+				}
+			}
+		}
+	}
```

Comment:
I think it'd be useful to either comment above this code block to the effect of `// Remove any existing Tink worker containers from previous runs` or put it in it own function.

Reply:
sure thing, will update. thanks